Automattic/wp-calypso
# FSE: Add domain-suggestions data store

Add `automattic/domains/suggestions` store script to FSE. This is a prerequisite for adding a domain picker in WP Admin.

See pbAok1-TR-p2

#### Changes proposed in this Pull Request

This PR adds a shared script to FSE. The script is called `a8c-fse-common-data-stores`. The script is built as part of the `common` bundle.

This necessitated a few changes to the webpack build:
- Allow multiple `source` arguments.
- Add special handling to webpack dependency extraction plugin.

In addition, a TypeScript definition has been added to the project to expose this "module". It will allow `import 'a8c-fse-common-data-stores';`.

The script is _registered_ in WordPress as part of the common bundle, but only enqueued as a script dependency when necessary.

This is important because these data stores should only be registered once (they should not be bundled in multiple scripts) and therefore should take advantage of this WordPress mechanism for sharing.

An alternative would be to rely on webpack dynamic imports, however that might require a major overhaul of the build system at this time because each package is built independently.

This is a first pass, scripts are only useful for their side effects. Further work would allow import/export from shared scripts if that were desired.

The only registered data store is `automattic/domains/suggestions` at this time.

#### Testing instructions

* Build and test with FSE - D43724-code can be used for easy testing.
* Sandbox `example.wordpress.com`
* Ensure that standard modules are externalized as before. You should see code like this in the build output:

```
/***/ "@wordpress/data":
/*!***************************************!*\
  !*** external {"this":["wp","data"]} ***!
  \***************************************/
/*! no static exports found */
/***/ (function(module, exports) {

(function() { module.exports = this["wp"]["data"]; }());

/***/ }),
```

* Add `import 'a8c-fse-common-data-stores';` to a module and ensure that `a8c-fse-common-data-stores` is added as a script dependency and causes the script to be enqueued.
* Test the data store, e.g.:
```js
wp.data.select( 'automattic/domains/suggestions' ).getCategories()
```

Note this will return `undefined` until it has resolved.
